### DISCUSSION TITLE
Graph panel: Ability to stack even with null values

### DISCUSSION THREAD

**Discussion by mplm17** [grafana/grafana#38298] [Dashboards, Panels & Visualizations]
**What would you like to be added**: Currently, in graph panel with multiple stacked metrics, if one metric sometimes have a null value that we want to ignore, we have the ability to set 'null value' parameter to 'connected'. The null value is so ignored and the graph looks like there is no null value. But, the others metrics do not take care about the 'connected' thing, here is how it does look like: 

![image](https://user-images.githubusercontent.com/39955446/121669963-3c044d80-caad-11eb-870a-1b33c457bccf.png)

It would be great to have some form of continuity on the others metrics too.

**Why is this needed**: I think this would be more readable.

Thanks

**Comment by torkelo** [comment]
Looks like that other metric does not have a null value but a zero value?

**Comment by torkelo** [comment]
Ah I see now what you mean

**Comment by mplm17** [comment]
> Looks like that other metric does not have a null value but a zero value?

Yes exactly, only the 'green' metric does have null values

**Comment by torkelo** [comment]
There is no easy way to solve this other than interpolation, or a group by time that is larger than the gaps

**Comment by mplm17** [comment]
Is this something that I can do by myself? Or something that could be done by grafana?

**Comment by 3DDario** [comment]
I would like to add that this was the default behavior in Grafana 7.3.10 and earlier. One of the many graph panel changes implemented between Grafana 7.3.10 and Grafana 7.4.0 must have broken the ability to connect null values when stacking is enabled.

I have attached two screenshots. The first one was generated with Grafana 7.3.10 and the second one with Grafana 7.4.0. The dataset, the time range and the entities between the two screenshots are the same. As you can see in my case the two entities don’t have the same update period so the problem becomes more noticeable.

**Grafana 7.3.10:**
![Grafana 7.3.10](https://i.imgur.com/mVfDWeB.png)

**Grafana 7.4.0:**
![Grafana 7.4.0](https://i.imgur.com/wqcG0GY.png)

**Comment by torkelo** [comment]
The stacking of null values required interpolating missing data points, and it caused numerous floating bars & and strange bugs. Not sure we will add that back. Is there another way to solve this using a transform?

**Comment by ploquets** [comment]
> or a group by time that is larger than the gaps

![image](https://user-images.githubusercontent.com/5149178/124684278-23693680-dea5-11eb-961b-5c1298bd3519.png)

**Comment by torkelo** [comment]
yes a group by time transform is one solution to this that I have though of

**Comment by ploquets** [comment]
> yes a group by time transform is one solution to this that I have though of

![image](https://user-images.githubusercontent.com/5149178/124833256-4a327600-df54-11eb-96ec-565425c12131.png)

When I group by time, I see **No graphable fields**

**Comment by ploquets** [comment]
If I change functions on each query to groupBy, avg, 5m

Then I start to see some changing
![image](https://user-images.githubusercontent.com/5149178/124833791-1277fe00-df55-11eb-9411-f159d56ccb11.png)

But, I would like to separate Upload / Download, like Upload being negative 
Tried to put Scale (-1) on those upload queries, but, even with the legend showing negative, doesn't seem to stack as negative.

![image](https://user-images.githubusercontent.com/5149178/124833602-d47ada00-df54-11eb-83e3-9cc56c7a6e89.png)

How should I do?

**Reply by jolcese** [reply]
How do do this?

**Comment by amarant** [comment]
It would be nice to have a last observation carried forward (locf) option to stack timeseries with different timestamps, because for now it looks like a regression,

**Comment by leeoniya** [comment]
i'm currently putting some groundwork into the underlying charting library (https://github.com/leeoniya/uPlot/commit/3bfbdbe36c39abe0bc544650c9ee528979936b61) to make this more tractable to implement in Grafana.

the current state of things is that uPlot will only render from raw data, so if we do any kind of interpolation for stacked and/or mis-aligned datasets, we need to insert fake datapoints, which will then show up in point rendering and during hover, tooltip, and legend interactions. this is not something i want, since it looks like there was real data when there actually wasn't.

the next uPlot update will allow each pathbuilder to pre-process the data that is only used for the canvas draw, so we can decouple the raw data from the rendered representations, which opens up the possibility to stack and align without introducing misleading filler data.

though i'll use this opportunity to voice my general distaste for using stacked line/area visualizations for analyzing timeseries trends: https://leeoniya.github.io/uPlot/demos/stacked-series.html


https://user-images.githubusercontent.com/43234/125665699-90051d81-f79f-4773-884d-a76c35bc4581.mp4

**Comment by amarant** [comment]
Thanks for your work, and your concerns, but in some situation there is no alternative, and with energy data we know the highly variable source from the constant ones, so we can stack them with an order to minimise the cumulative variability effects, also we need stacking to compare cumulative sources with the demand.

**Comment by ploquets** [comment]
Well, for me, the workaround for now was to:

Create two graph itens for those I want to stack reverse

- One Negative (grupby and scale -1)
- One Positve (only grupby)

For these, I created an override rule to make them use same color.
Then arrange items as negatives first and positives after

Positive stacking always start from the last value stacked, so, I need to make it zero again (by adding same values negative and positive) and the others positives started to stack from zero.

![image](https://user-images.githubusercontent.com/5149178/125860492-ab48dbc1-e3bb-411c-84df-e90e4afd893a.png)

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by gelicia** [comment]
Closing due to deprecation of graph panel. Please create a [new issue ticket](https://github.com/grafana/grafana/issues) describing the feature request.
